Minimal exception handling extension

dhil opened the pull request:
This patch extends the interpreter with a minimal exception handling facility. It is minimal in the sense that it supports only a single exception which can be thrown and caught. I think this should be sufficient to provide a base for further experiments.
    
The patch is implemented on top of the latest reference-types proposal, though in retrospect it seems to be unnecessary as with this minimal design there is no notion of exception types, and hence no need for reference types.

dhil committed:
Minimal exception handling extension

This patch extends the interpreter with a minimal exception handling
facility. It is minimal in the sense that it supports only a single
exception which can be thrown and caught.

dhil commented:
I can potentially disentangle the exception handling extension from the reference-types commits.

rossberg commented:
Yes, I'd suggest separating the upstream merge from the exception PR. I just created #2, so maybe it's sufficient if you approve that and then this PR automatically adopts that as baseline -- though there might be some merge conflicts.

dhil commented:
Sounds good to me! I will give it a try.

dhil commented:
@rossberg I have updated the patch now. It is now a single commit on top of master with typed-func-refs merged in.

rossberg commented on interpreter/syntax/free.ml:
```diff
@@ -123,6 +123,8 @@ let rec instr (e : instr) =
     memories zero
   | MemoryInit x -> memories zero ++ datas (idx x)
   | DataDrop x -> datas (idx x)
+  | Try (_, es1, es2) -> block es1 ++ block es2
```
```suggestion
  | Try (bt, es1, es2) -> block_type bt ++ block es1 ++ block es2
```

rossberg commented on interpreter/text/arrange.ml:
```diff
@@ -288,6 +288,10 @@ let rec instr e =
     | Convert op -> cvtop op, []
+    | Try (bt, es1, es2) ->
+      "try", block_type bt @
+               [Node ("then", list instr es1); Node ("catch", list instr es2)]
```
```suggestion
        [Node ("do", list instr es1); Node ("catch", list instr es2)]
```

rossberg commented on interpreter/text/parser.mly:
```diff
@@ -692,6 +698,43 @@ call_expr_results :
+      let out' = snd $3 c in
+      FuncType (ins, out' @ out), es }
+try_ :
+ | LPAR THEN instr_list RPAR LPAR CATCH instr_list RPAR
```
Current proposal uses `do` for this. (May need a new token)
```suggestion
 | LPAR DO instr_list RPAR LPAR CATCH instr_list RPAR
```

rossberg commented on interpreter/binary/encode.ml:
```diff
@@ -428,6 +428,8 @@ let encode m =
       | Convert (F64 F64Op.DemoteF64) -> assert false
       | Convert (F64 F64Op.ReinterpretInt) -> op 0xbf
 
+      | Try _ | Throw -> failwith "encode.ml: not yet implemented"
```
```suggestion
      | Try _ | Throw -> failwith "encode: not yet implemented"
```

rossberg commented on interpreter/exec/eval.ml:
```diff
@@ -525,7 +528,17 @@ let rec step (c : config) : config =
+        let FuncType (ts1, ts2) = block_type c.frame.inst bt e.at in
+        let n1 = List.length ts1 in
+        let n2 = List.length ts2 in
+        let args, vs' = take n1 vs e.at, drop n1 vs e.at in
```
```suggestion
        let args, vs' = split n1 vs e.at in
```

rossberg commented on interpreter/exec/eval.ml:
```diff
@@ -616,6 +635,20 @@ let rec step (c : config) : config =
+    | Catch (_, _, (_, ({it = Trapping _ | Breaking _ | Returning _; _} as e) :: _)), vs ->
+      vs, [e]
+
+    | Catch (n, es1, (_, { it = Throwing; at} :: _)), vs ->
```
```suggestion
    | Catch (n, es1, (_, {it = Throwing; _} :: _)), vs ->
```

rossberg commented on interpreter/exec/eval.ml:
```diff
@@ -616,6 +635,20 @@ let rec step (c : config) : config =
+      vs, [e]
+
+    | Catch (n, es1, (_, { it = Throwing; at} :: _)), vs ->
+      let exn = [] in
+      vs, [Label (n, [], (exn, List.map plain es1)) @@ e.at]
```
```suggestion
      vs, [Label (n, [], ([], List.map plain es1)) @@ e.at]
```

dhil committed:
Update interpreter/syntax/free.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Update interpreter/text/arrange.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Update interpreter/text/parser.mly

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Update interpreter/exec/eval.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Update interpreter/exec/eval.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Update interpreter/binary/encode.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Update interpreter/exec/eval.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil commented:
Thanks!

dhil committed:
Declare DO as a token.

dhil committed:
Add lexical rules for try/catch/do/throw.

dhil committed:
Add a few test cases.

dhil commented:
@rossberg I have added a few test cases now (and a little bit of infrastructure to support testing, e.g. `assert_uncaught`).

dhil committed:
Implement binary encoding and decoding.

The opcodes for try/catch/throw are taken from the provisional
exception-handling proposal.

dhil commented:
Compiles and passes the test suite now. 

To make it pass the test suite I had to implement encode/decode for exceptions and their handlers. I have used the opcodes from the current iteration of the exception handling proposal.

rossberg commented on interpreter/script/js.ml:
```diff
@@ -529,6 +529,8 @@ let of_assertion mods ass =
   | AssertExhaustion (act, _) ->
     of_assertion' mods act "assert_exhaustion" [] None
+  | AssertUncaught (act, _) ->
+    of_assertion' mods act "assert_uncaught" [] None
```
You'll need to add that function to the JS harness above.

rossberg commented on interpreter/script/run.ml:
```diff
@@ -446,6 +447,13 @@ let run_assertion ass =
     )
 
+  | AssertUncaught (act, re) ->
+    trace ("Asserting trap...");
```
```suggestion
    trace ("Asserting exception...");
```

rossberg commented on interpreter/script/run.ml:
```diff
@@ -446,6 +447,13 @@ let run_assertion ass =
+    trace ("Asserting trap...");
+    (match run_action act with
+    | exception Eval.Uncaught (_, msg) -> assert_message ass.at "runtime" msg re
+    | _ -> Assert.error ass.at "expected runtime error"
```
```suggestion
    | _ -> Assert.error ass.at "expected exception"
```

rossberg commented on interpreter/text/lexer.mll:
```diff
@@ -396,6 +396,7 @@ rule token = parse
   | "catch" { CATCH }
   | "do" { DO }
   | "throw" { THROW }
+  | "assert_uncaught" { ASSERT_UNCAUGHT }
```
Nit: group with other assertions.

rossberg commented on interpreter/text/parser.mly:
```diff
@@ -228,7 +228,7 @@ let inline_func_type_explicit (c : context) x ft at =
 %token INPUT OUTPUT
 %token EOF
-%token TRY CATCH DO THROW
+%token TRY CATCH DO THROW ASSERT_UNCAUGHT
```
Nit: group with other assertions

rossberg commented on interpreter/binary/decode.ml:
```diff
@@ -270,7 +270,17 @@ let rec instr s =
   | 0x05 -> error s pos "misplaced ELSE opcode"
-  | 0x06| 0x07 | 0x08 | 0x09 | 0x0a as b -> illegal s pos b
+  | 0x06 ->
+     let bt = block_type s in
```
Nit: indent -= 1

rossberg reviewed (commented):
Nice! A couple more comments about that.

dhil committed:
Update interpreter/script/run.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Update interpreter/script/run.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>

dhil committed:
Address @rossberg's comments.

dhil committed:
Adds missing success reduction for `catch`.

Extends test suite too.

dhil commented:
@rossberg I had forgotten to add a reduction rule for the success-case when the body of `try` raises no exceptions. My latest commit adds that rule. With this, I think the patch is complete.

rossberg commented on interpreter/script/js.ml:
```diff
@@ -189,6 +189,13 @@ function assert_return(action, ...expected) {
+
+function assert_uncaught(action) {
+  try { action() } catch (e) {
+    if (e instanceof WebAssembly.RuntimeError) return;
```
```suggestion
    if (!(e instanceof WebAssembly.RuntimeError)) return;
```

dhil committed:
Update interpreter/script/js.ml

Co-authored-by: Andreas Rossberg <rossberg@mpi-sws.org>
